File: LeetCode/Study Plan/Practice Problems/Hard/Edit Distance/Java/editDistance.java
class Solution {
    int[][] dp;
    int n;
    int m;

    public int minDistance(String word1, String word2) {
        n = word1.length();
        m = word2.length();
        dp = new int[n][m];
        
        for (int i=0; i<n; i++) {
            Arrays.fill(dp[i], -1);
        }

        return editDistance(0, 0, word1, word2);
    }

    public int editDistance(int i, int j, String word1, String word2) {
        if (i == n) return m-j;
        if (j == m) return n-i;
        if (dp[i][j] != -1) return dp[i][j];

        if (word1.charAt(i) == word2.charAt(j)) dp[i][j] = editDistance(i+1, j+1, word1, word2);
        else {
            int insert = 1 + editDistance(i, j+1, word1, word2);
            int remove = 1 + editDistance(i+1, j, word1, word2);
            int replace = 1 + editDistance(i+1, j+1, word1, word2);
            dp[i][j] = Math.min(insert, Math.min(remove, replace));
        }

        return dp[i][j];
    }
}

File: LeetCode/Study Plan/Practice Problems/Hard/Edit Distance/Go/editDistance.go
package main

import "fmt"

//------------------------brute force try to generate all possibilities---------------------
func minDistanceBruteForce(word1 string, word2 string) int {
	return editDistanceBruteForce(0, 0, word1, word2)
}

func editDistanceBruteForce(i, j int, word1, word2 string) int {
	if i == len(word1) {
		return len(word2[j:])
	}

	if j == len(word2) {
		return len(word1[i:])
	}

	if word1[i] != word2[j] {
		insert := 1 + editDistanceBruteForce(i, j+1, word1, word2)
		remove := 1 + editDistanceBruteForce(i+1, j, word1, word2)
		replace := 1 + editDistanceBruteForce(i+1, j+1, word1, word2)
		return min(insert, min(remove, replace))
	} else {
		return editDistanceBruteForce(i+1, j+1, word1, word2)
	}
}

//------------------------brute force try to generate all possibilities---------------------

//--------------------------------memoization----------------------------
func minDistance(word1, word2 string) int {
	n := len(word1)
	m := len(word2)

	dp := make([][]int, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]int, m)
		for j := 0; j < m; j++ {
			dp[i][j] = -1
		}
	}

	return editDistance(0, 0, word1, word2, &dp)
}

func editDistance(i, j int, word1, word2 string, dp *[][]int) int {
	if i == len(word1) {
		return len(word2) - j
	}

	if j == len(word2) {
		return len(word1) - i
	}

	if (*dp)[i][j] != -1 {
		return (*dp)[i][j]
	}

	if word1[i] != word2[j] {
		insert := 1 + editDistance(i, j+1, word1, word2, dp)
		remove := 1 + editDistance(i+1, j, word1, word2, dp)
		replace := 1 + editDistance(i+1, j+1, word1, word2, dp)
		(*dp)[i][j] = min(insert, min(remove, replace))
	} else {
		(*dp)[i][j] = editDistance(i+1, j+1, word1, word2, dp)
	}
	return (*dp)[i][j]
}

//--------------------------------memoization----------------------------

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func main() {
	fmt.Println(minDistance("horse", "ros"))
}
